mlir/include/mlir/Dialect/MIOpenOps/MIOpenCPP.h

```diff
@@ -91,51 +91,325 @@ T integer_least_multiple(T x, T y)
+      bool &input2GemmKVectorizable) {
+    // Vectorizable flag is opposite between forwad and bwd_data
+    if (opType == mlir::miopen::ConvOpType::Conv2DOpType) {
+      // For input tensor.
+      // When C is the fastest changing dimension,
+      // gemmK dimension is vectorizable, gemmN is not, and vice versa.
+      // Vectorization width depending on length of C.
+      if (dimIndexVal["ci"].first == 3) {
+        input2GemmKVectorizable = true;
+      } else {
+        input2GemmKVectorizable = false;
+      }
+    } else if (opType == mlir::miopen::ConvOpType::Conv2DBwdDataOpType) {
+      // For output tensor.
+      // When K is the fastest changing dimension(3),
+      // gemmK dimension is vectorizable, gemmN is not, and vice versa.
+      // Vectorization width depending on length of K.
+      if (dimIndexVal["ko"].first == 3) {
+        input2GemmKVectorizable = true;
+      } else {
+        input2GemmKVectorizable = false;
+      }
+    }
+  }
+
+  static void
+  obtainNCHWVecLen(llvm::StringMap<std::pair<size_t, int64_t>> &dimIndexVal,
+                   int64_t &vecLen) {
+    if (dimIndexVal["ni"].first == 3) {
+      vecLen = dimIndexVal["ni"].second;
+    } else if (dimIndexVal["ci"].first == 3) {
+      vecLen = dimIndexVal["ci"].second;
+    } else {
+      // Not vectorizable
+      // TODO(optimize): For 1x1, stride 1, pad 0 conv, vecLen is hi * wi
+      vecLen = 1;
+    }
+  }
+  static void
+  obtainNKHWVecLen(llvm::StringMap<std::pair<size_t, int64_t>> &dimIndexVal,
+                   int64_t &vecLen) {
+    if (dimIndexVal["ko"].first == 3) {
+      vecLen = dimIndexVal["ko"].second;
+    } else if (dimIndexVal["ko"].first == 0) {
+      // dimKO is the lowest changing dimension, which means dimN/dimHo/dimWo
+      vecLen = dimIndexVal["no"].second * dimIndexVal["ho"].second *
+               dimIndexVal["wo"].second;
+    } else if (dimIndexVal["ko"].first == 1) {
+      // Ko's position is at 1, vectorization legnth is last two dimensions
+      if (dimIndexVal["no"].first == 0) {
+        vecLen = dimIndexVal["ho"].second * dimIndexVal["wo"].second;
+      } else if (dimIndexVal["ho"].first == 0) {
+        vecLen = dimIndexVal["no"].second * dimIndexVal["wo"].second;
+      } else {
+        vecLen = dimIndexVal["no"].second * dimIndexVal["ho"].second;
+      }
+    } else {
+      // K's position is 2, vectorization legnth is last dimension
+      if (dimIndexVal["no"].first == 3) {
+        vecLen = dimIndexVal["no"].second;
+      } else if (dimIndexVal["ho"].first == 3) {
+        vecLen = dimIndexVal["ho"].second;
+      } else {
+        vecLen = dimIndexVal["wo"].second;
+      }
+    }
+  }
+
+  static void
+  obtainInput2VecLen(mlir::miopen::ConvOpType opType,
+                     llvm::StringMap<std::pair<size_t, int64_t>> &dimIndexVal,
+                     int64_t &vecLen) {
+    if (opType == mlir::miopen::ConvOpType::Conv2DOpType) {
```

Comment:
comment here needed to describe why we are only looking for a specific layout, but not other layouts.

Reply:
This is probably a variable naming issue. The function name is `obtainNCHWVecLen`. That makes it easy for the reader to think it will only fit for one `NCHW` layout. In reality, 

- It is `NCHW` because the second input gemm in forward pass will always be a variation of `NCHW`
- this function handles all combinations of `NCHW`. 

I'll probably ended up adding a comment at the declaration of each of the function that has layout letters in them clarifying that this does not stand for one particular layout sequence.

Reply:
Done. 

Instead of naming it with a specific layout, I renamed the function according to the representation of gemm at the specific position. For example:

In forward: `Filter x Input = Output`. To represent the second gemm, I now use `obtainInputVecLen` instead.